lib/macs/cmac.dart

```diff
@@ -0,0 +1,300 @@
+ * </p><p>
+ * It supports 128- or 64-bits block ciphers, with any key size, and returns
+ * a MAC with dimension less or equal to the block size of the underlying 
+ * cipher.
+ * </p>
+ */
+class CMac extends BaseMac {
+  static final FactoryConfig FACTORY_CONFIG = new DynamicFactoryConfig.suffix(
+      Mac,
+      "/CMAC",
+      (_, final Match match) => () {
+            BlockCipher cipher = new BlockCipher(match.group(1));
+            return new CMac(cipher, cipher.blockSize * 8);
+          });
+
+  Uint8List _poly;
+  Uint8List _ZEROES;
+
+  Uint8List _mac;
+
+  Uint8List _buf;
+  int _bufOff;
+  final BlockCipher _cipher;
+
+  final int _macSize;
+
+  Uint8List _Lu, _Lu2;
+
+  ParametersWithIV _params;
+
+  /**
+     * create a standard MAC based on a CBC block cipher (64 or 128 bit block).
+     * This will produce an authentication code the length of the block size
+     * of the cipher.
+     *
+     * @param cipher the cipher to be used as the basis of the MAC generation.
+     */
+//    CMac(BlockCipher cipher) : CMac(this._cipher, _cipher.blockSize * 8);
+
+  /**
+     * create a standard MAC based on a block cipher with the size of the
+     * MAC been given in bits.
+     * <p>
+     * Note: the size of the MAC must be at least 24 bits (FIPS Publication 81),
+     * or 16 bits if being used as a data authenticator (FIPS Publication 113),
+     * and in general should be less than the size of the block cipher as it reduces
+     * the chance of an exhaustive attack (see Handbook of Applied Cryptography).
+     *
+     * @param cipher        the cipher to be used as the basis of the MAC generation.
+     * @param macSizeInBits the size of the MAC in bits, must be a multiple of 8 and &lt;= 128.
+     */
+  CMac(BlockCipher cipher, int macSizeInBits)
+      : this._macSize = macSizeInBits ~/ 8,
+        this._cipher = new CBCBlockCipher(cipher) {
+//  CMac(BlockCipher cipher)
+//      : this._macSize = cipher.blockSize,
+//        this._cipher = new CBCBlockCipher(cipher) {
+    //int macSizeInBits = cipher.blockSize * 8;
+
+    if ((macSizeInBits % 8) != 0) {
+      throw new ArgumentError("MAC size must be multiple of 8");
+    }
+
+    if (macSizeInBits > (_cipher.blockSize * 8)) {
+      throw new ArgumentError(
+          "MAC size must be less or equal to ${_cipher.blockSize * 8}");
+    }
+
+    _poly = lookupPoly(cipher.blockSize);
+
+    _mac = new Uint8List(cipher.blockSize);
+
+    _buf = new Uint8List(cipher.blockSize);
+
+    _ZEROES = new Uint8List(cipher.blockSize);
+
+    _bufOff = 0;
+  }
+
+  @override
+  String get algorithmName {
+    String blockCipherAlgorithmName = _cipher.algorithmName.split("/").first;
+    return "${blockCipherAlgorithmName}/CMAC";
+  }
+
+  static int shiftLeft(Uint8List block, Uint8List output) {
+    int i = block.length;
+    int bit = 0;
+    while (--i >= 0) {
+      int b = block[i] & 0xff;
+      output[i] = ((b << 1) | bit);
+      bit = (b >> 7) & 1;
```

Comment:
This was a triple shift right in BouncyCastle (`>>>`). I hope that doesn't matter.

Reply:
I think I have also struggled with that. I don't think there is a meaningful difference in this case.